drop-transfer/src/ws/events.rs

```diff
@@ -250,23 +270,41 @@ impl FileEventTx<IncomingTransfer> {
 
 impl FileEventTx<OutgoingTransfer> {
     pub async fn start(&self, offset: u64) {
-        self.start_inner(crate::Event::FileUploadStarted(
-            self.xfer.clone(),
-            self.file_id.clone(),
-            offset,
-        ))
-        .await
+        let events = [
+            crate::Event::FileUploadStarted(self.xfer.clone(), self.file_id.clone(), offset),
+            crate::Event::FileUploadProgress(self.xfer.clone(), self.file_id.clone(), offset),
+        ];
+
+        self.start_inner(events).await
```

Comment:
So the reason is as follows:
At first, I tried to implement it the easy way, meaning to shoot the events in the spawned tasks. It failed the test cases spectacularly.
Note that the previous behavior was to shoot a `Started` event and only then spawn the task. That was done to guarantee ale the `Progress` events arrived after the `Started` one.
Why was that? Because the `Progress` event is generated in the other thread, as a result of peer message, when I moved the `Started` inside the task, then couple of `Progress` messages were dropped (these are dropped in the `ws/event.rs`) before `Started` was emitted. It might be fine however a lot of tests depend on receiving `Progress` events in an exact order, so probably apps rely on that as well.
So I planned to go for the initial check outside the task, and emit `Started` or `Throttled` as a result, and then move the eventual waiting for the permit to the task with the `Started` event. That does not fix the issue fully, but for the majority of cases, and as I mentioned it's not an issue per se, so we might be good with that compromise.
I hope I've explained it clearly.

PS You do have a point here that it shouldn't be a double event in case the first permit acquisition succeeds, I will fix this one.

Reply:
I've pushed the PS change

Reply:
Thanks a lot for the explanation, it does make sense and good point on sending the `Progress` event on start to make up for dropped ones before

Reply:
This explanation doesn't make it fully clear to me why we need to compensate the progress events.. Shouldn't it be like this:

- 1) spawn a task
- 2) inside the task await for permit
- 3) if we fail to receive a permit, spawn throttled(if we haven't yet)
- 4) loop 2
- 5) if we succeed, emit start event and enter the main loop to emit progress events

Actually it raises a question - if we don't loop, then we will emit throttle event and will not upload again until manually pinged via `norddrop_network_refresh()` API call

Reply:
@LukasPukenis the `refresh()` call has nothing to do with the file upload task, it's on a much higher level. The file upload task is within one turn of a loop so to speak.
That said, you don't wait for the permit in a loop. You just wait for it and then proceed. The problem though is that the `Progress` event is emitted elsewhere, as a reaction to a peer message, for which you don't have control inside the upload task. That is why some of them are dropped and we need to make up at least one of them.
As I mentioned it's technically OK to drop some progress events but since our tests rely on them not being dropped, I decided to not break the tests and potential apps.

So to answer your question, the `refresh()` API is unrelated to the throttling problem